### ISSUE TITLE
Question about the usage of NTXentLoss along with CrossBatchMemory

### ISSUE THREAD

**Issue by HaominJiang** [KevinMusgrave/pytorch-metric-learning#609]
Hi Kevin, @KevinMusgrave I have tried to apply NTXentLoss with CrossBatchMemory but the average precision of validation set always increased fast during warm-up and kept dropping after that. I'm wondering whether my usage of these loss functions was wrong.

![question](https://user-images.githubusercontent.com/74887068/231399941-4bcf8f6b-304b-4628-b409-b1dc1bbc2a87.png)

Here are my settings:

- Model structure: a pretrained ViT-B/32 backbone with a 2-layers MLP
- Dataset: generate positive pairs from the same image by different augmentation parameters, which is similar to SimCLR
- Optimizer: AdamW
- LR sheduler: warmup + cosine LR decay
- Loss function configuration:
```python
loss_fn = losses.NTXentLoss(temperature=0.07)
loss_fn = losses.CrossBatchMemory(loss_fn, embedding_size=128, memory_size=2048)
```
- Loss calculation in one epoch training:
Each image is augmented by different parameters twice and each batch may have images from the same class. Images and labels in one batch look like (assume anchor1 and anchor5 have the same label in this batch):
**images**: `[anchor1, positive1, anchor2, positive2, anchor3, positive3, anchor4, positive4, anchor5, positive5...]`
**labels**: `[0, 0, 2, 2, 5, 5, 3, 3, 0, 0...]`
```
embeddings = model(images)
loss = loss_fn(embeddings, labels)
```

Is there anything wrong in my usage of the NTXentLoss with CrossBatchMemory that related to the drop of AP? I have tried many LR and weight decay settings but they all failed to stop the decrease of AP.

Looking forward to and thanks for your feedback~

**Comment by HaominJiang** [comment]
Yep, I used the CrossBatchMemory wrapped NTXentLoss during the whole training process, including the warm-up stage. Should I only use pure NTXentLoss during the warm-up?
For labels, they are consistent across batches. Could I compute the loss by just feeding the augmented image pairs and their labels to the loss function as shown in the above codes?

**Comment by KevinMusgrave** [comment]
> Yep, I used the CrossBatchMemory wrapped NTXentLoss during the whole training process, including the warm-up stage. 

What's your batch size? Do you process more than 2048 examples during warm-up? (2048 because that's the `memory_size`).

> For labels, they are consistent across batches. Could I compute the loss by just feeding the augmented image pairs and their labels to the loss function as shown in the above codes?

Yeah that should work fine.

It seems like NTXentLoss + CrossBatchMemory is working. Otherwise why would the accuracy go up during the warm-up stage?

I suppose you could try NTXentLoss by itself, and see if you get the same behavior.

**Comment by KevinMusgrave** [comment]
> My batch size is 256, which is much smaller than 2048. So in this configuration, could I use the CrossBatchMemory wrapped NTXentLoss in warm-up stage?

That's fine. I was thinking maybe there is a bug that occurs only after the queue fills up. But in your case the queue will fill up very quickly during the warmup stage, so there doesn't seem to be an issue with the queue.

> Another question is that I found a similar issue [#573 (comment)](https://github.com/KevinMusgrave/pytorch-metric-learning/issues/573#issuecomment-1385452570) and you have suggested using half-batch for loss computation. What is the difference between the half-batch method and mine (directly feeding labels to loss computation)?

In that comment I was constructing the labels to make the subsequent operations easier. It's not necessary in your case.

**Comment by HaominJiang** [comment]
I tried 3 loss function settings with the same other hyper parameters, the test AP did not keep dropping only when NTXentLoss worked alone. It seemed that the performance of NTXentLoss can be decreased by CrossBatchMemory. Do you have any ideas or suggestions?

**Green Line:** Apply CrossBatchMemory wrapped NTXentLoss in the whole training progress including warm-up stage
**Blue Line:** Apply pure NTXentLoss on warm-up stage and CrossBatchMemory wrapped NTXentLoss after that
**Orange Line:** Only apply NTXentLoss

![comparsion](https://user-images.githubusercontent.com/74887068/232183419-411f8e34-7539-4ed4-8a51-fc1dffad4e80.png)

**Comment by KevinMusgrave** [comment]
Interesting results! I'm not sure what causes this behavior though 🤔

**Comment by KevinMusgrave** [comment]
The only thoughts I have are:

- Maybe the features in memory are too old. But then why does it seem to work well initially, for the first 5 epochs?
- When features are added to the queue, they are detached from the autograd graph. For a positive and negative pair (a1, p) and (a2, n), only a1 and a2 are attached to the autograd graph. In contrast, when you use the "pure" loss function, all embeddings (a1, p, a2, n) are attached to the autograd graph.

One way to fix the 2nd point could be to compute both the "pure" loss and the xbm loss:

```python
loss_pure = loss_fn_pure(embeddings, labels)
loss_xbm = loss_fn_xbm(embeddings, labels)
loss = loss_pure + loss_xbm
```

But at this point I'm sort of guessing.

**Comment by HaominJiang** [comment]
Thanks! Maybe I should use a smaller memory size.

Another question is that the `loss_xbm` in my code is CrossBatchMemory wrapped NTXentLoss, doesn't it automatically compute the `xbm loss + loss passed to the function`?

**Comment by HaominJiang** [comment]
I tried more loss function settings and found
1. SupConLoss performs better than NTXentLoss, perhaps because our data are all labeled
2. CrossBatchMemory may decrease the accuracy of SupConLoss (as well as NTXentLoss, shown in conversations before). 
> Green Line: Apply CrossBatchMemory wrapped NTXentLoss in the whole training progress including warm-up stage
Blue Line: Apply pure NTXentLoss on warm-up stage and CrossBatchMemory wrapped NTXentLoss after that
Orange Line: Only apply NTXentLoss

The result may be consistent with their claim that xbm benefits to the contrastive loss ([the traditional one](https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#contrastiveloss)) most in the paper https://arxiv.org/pdf/1912.06798.pdf

**Maybe I should not apply XBM to NTXentLoss/SupConLoss.**

Here are the configs in my experiments:
**Orange Line:** Only apply NTXentLoss
**Light Blue Line:** Only apply SupConLoss 
**Pink Line:** Apply SupConLoss with CrossBatchMemory (memory size=1024, loss during warm-up=only supconloss, loss after warm-up = supconloss + xbmloss)

![Supcon](https://user-images.githubusercontent.com/74887068/232965384-75c08f6b-bf7d-449f-8f40-8e82c0f3ae52.png)

**Comment by KevinMusgrave** [comment]
Thanks for posting your results!

It's interesting that SupConLoss performs better. I suppose one potential cause is the different softmax temperature:
- SupConLoss, t=0.1
- NTXentLoss, t=0.07

**Comment by HaominJiang** [comment]
In all the above experiments, the temperature parameter was kept constant, which means I apply temperature=0.07 for both NTXentLoss and SupConLoss. Besides, I kept all the other hyper papameters like LR, weight decay unchanged and **only adjust the form of the loss function.**

**Comment by KevinMusgrave** [comment]
Ah I see, so SupConLoss really does perform significantly better than NTXentLoss.

**Comment by HaominJiang** [comment]
Yep. More precisely, my guess is that in the case of having well-labeled training data, SupConLoss has more advantages than NTXentLoss.
Besides, both of them may not be benefited from CrossBatchMemory.
